Fabric.Identity.API/Services/ClaimsService.cs

```diff
@@ -0,0 +1,147 @@
+                                  .Value;
+            }
+
+            if(subjectId == null)
+            {
+                subjectId = user.SubjectId;
+            }
+
+            return subjectId;
+        }
+
+        private Claim[] GenerateAdditionalClaims(List<Claim> previousClaims)
+        {
+            var additionalClaims = new List<Claim>();
+            //if the external system sent a session id claim, copy it over
+            var sid = previousClaims.FirstOrDefault(x => x.Type == JwtClaimTypes.SessionId);
+            if (sid != null)
+            {
+                additionalClaims.Add(new Claim(JwtClaimTypes.SessionId, sid.Value));
+            }
+
+            //if the external provider issues groups claims, copy it over
+            var groupClaims = previousClaims.Where(c => c.Type == "groups").ToList();
+            if (groupClaims.Any())
+            {
+                additionalClaims.AddRange(groupClaims);
+            }
+
+            return additionalClaims.ToArray();
+        }
+
+        private AuthenticationProperties GenerateAuthenticationProperties(AuthenticateInfo info)
+        {
+            //if the external provider issued an id_token, we'll keep it for signout
+            AuthenticationProperties props = null;
+            var id_token = info.Properties.GetTokenValue("id_token");
+            if (id_token != null)
+            {
+                props = new AuthenticationProperties();
+                props.StoreTokens(new[] { new AuthenticationToken { Name = "id_token", Value = id_token } });
+            }
+
+            return props;
+        }
+
+        private bool ValidateAzureAD(List<Claim> claims)
+        {
+            var issuerClaim = claims.FirstOrDefault(x => x.Type == JwtClaimTypes.Issuer);
+            if (issuerClaim == null)
+            {
+                throw new MissingIssuerClaimException(ExceptionMessageResources.MissingIssuerClaimMessage);
+            }
+
+            return !this._appConfiguration.AzureActiveDirectorySettings.IssuerWhiteList.Contains(issuerClaim.Issuer);
+        }
+
+        private Claim GetUserIdClaim(List<Claim> claims)
+        {
+            //try to determine the unique id of the external user - the most common claim type for that are the sub claim and the NameIdentifier
+            //depending on the external provider, some other claim type might be used
+            var userIdClaim = claims.FirstOrDefault(x => x.Type == JwtClaimTypes.Subject);
+            if (userIdClaim == null)
+            {
+                userIdClaim = claims.FirstOrDefault(x => x.Type == ClaimTypes.NameIdentifier);
+            }
+
+            if (userIdClaim == null)
+            {
+                throw new MissingUserClaimException(ExceptionMessageResources.MissingUserClaimMessage);
+            }
+
+            return userIdClaim;
+        }
+
+        private bool IsAzureEnabled(string schemaItem) => 

```

Comment:
This is really more than merely AzureAD being enabled, its that AzureAd is enabled, and the authentication scheme is azure for the current user - perhaps we should rename this one.